pennylane/transforms/decompositions/clifford_plus_t/cliffordt_transform.py

```diff
@@ -0,0 +1,368 @@
+from pennylane.transforms.core import transform
+from pennylane.tape import QuantumTape
+from pennylane.transforms.optimization import (
+    cancel_inverses,
+    commute_controlled,
+    merge_rotations,
+    remove_barrier,
+)
+from pennylane.transforms.optimization.optimization_utils import find_next_gate, _fuse_global_phases
+
+# Single qubits Clifford+T gates in PL
+_CLIFFORD_T_ONE_GATES = [
+    qml.Identity,
+    qml.PauliX,
+    qml.PauliY,
+    qml.PauliZ,
+    qml.Hadamard,
+    qml.S,
+    qml.SX,
+    qml.T,
+]
+
+# Two qubits Clifford+T gates in PL
+_CLIFFORD_T_TWO_GATES = [
+    qml.CNOT,
+    qml.CY,
+    qml.CZ,
+    qml.SWAP,
+    qml.ISWAP,
+]
+
+_PARAMETER_GATES = [qml.RX, qml.RY, qml.RZ, qml.Rot, qml.GlobalPhase]
+
+_CLIFFORD_T_GATES = _CLIFFORD_T_ONE_GATES + _CLIFFORD_T_TWO_GATES
+
+
+# pylint: disable= bare-except, unnecessary-lambda-assignment
+def check_clifford_op(op):
+    r"""Check if an operator is Clifford or not.
+
+    For a given unitary operator :math:`U` acting on :math:`N` qubits, this method checks that the
+    transformation :math:`UPU^{\dagger}` maps the Pauli tensor products :math:`P = {I, X, Y, Z}^{\otimes N}`
+    to Pauli tensor products with :math:`O(N \times 8^N)` time complexity when using naive matrix multiplication.
+
+    Args:
+        op: the operator that needs to be tested
+
+    Returns:
+        Bool that represents whether the provided operator is Clifford or not.
+    """
+
+    # Check if matrix can be calculated for the operator
+    if not op.has_matrix:
+        return False
+
+    # Compute the LCUs for the operator in Pauli basis
+    num_qubits = len(op.wires)
+    op = qml.map_wires(op, wire_map={wire: idx for idx, wire in enumerate(op.wires)})
+    op_matrix = qml.matrix(op, wire_order=range(num_qubits))
+    pauli_terms = qml.pauli_decompose(op_matrix, check_hermitian=False)
+
+    # Build Pauli tensor products P in Hamiltonian form
+    def pauli_group(x):
+        return [qml.Identity(x), qml.PauliX(x), qml.PauliY(x), qml.PauliZ(x)]
+
+    pauli_qubit = [
+        qml.prod(*pauli) for pauli in product(*(pauli_group(idx) for idx in range(num_qubits)))
+    ]
+    pauli_hams = [
+        qml.pauli.pauli_sentence(op).hamiltonian(wire_order=range(num_qubits)) for op in pauli_qubit
+    ]
+
+    # Perform U@P@U^\dagger and check if the result exists in set P
+    pauli_coves = []
+    for prod in product([pauli_terms], pauli_hams, [pauli_terms]):
+        # hopefully op_math.prod scales better than matrix multiplication, i.e., O((2^N)^3)
+        upu = qml.pauli.pauli_sentence(qml.prod(*prod))
+        upu.simplify()
+
+        upu2 = upu.hamiltonian(wire_order=range(num_qubits))
+        if not len(upu2.ops) == 1:  # Pauli sum always lie outisde set P
+            return False  # early stopping
+        if not isinstance(upu2.ops[0], qml.Identity):  # Identity always lie in set P
+            pauli_coves.append(any((qml.equal(upu2.ops[0], tm) for tm in pauli_hams)))
```

Comment:
Could we just perform all of these comparisons on pauli sentences themselves?  Maybe we need an "is close" comparison function for pauli sentences first?

Reply:
It should be possible but it would not be as straightforward as doing this. :)